## NB1

# Import the libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Set a nice style for the plots
sns.set_theme(style="whitegrid")

In [2]:
# Load the data set
df = pd.read_csv("C:/Projects/10_Stroke_risk_prediction/data/healthcare-dataset-stroke-data.csv")

In [3]:
# Display the first 5 rows
df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


# Inspect the dataset

In [4]:
# check data types and non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [5]:
# Check for missing values
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

In [6]:
# View statistical summaries of numericaL features
df.describe()

,id,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke
count,5110.000000,5110.000000,5110.000000,5110.000000,5110.000000,4909.000000,5110.000000
mean,36517.829354,43.226614,0.097456,0.054012,106.147677,28.893237,0.048728
std,21161.721625,22.612647,0.296607,0.226063,45.283560,7.854067,0.215320
min,67.000000,0.080000,0.000000,0.000000,55.120000,10.300000,0.000000
25%,17741.250000,25.000000,0.000000,0.000000,77.245000,23.500000,0.000000
50%,36932.000000,45.000000,0.000000,0.000000,91.885000,28.100000,0.000000
75%,54682.000000,61.000000,0.000000,0.000000,114.090000,33.100000,0.000000
max,72940.000000,82.000000,1.000000,1.000000,271.740000,97.600000,1.000000


# Data cleaning

In [7]:
# drop the id column
df = df.drop(columns=['id'])

In [8]:
## For missing values for bmi
# Check the percentage of missing values to confirm
missing_percentage = (df['bmi'].isnull().sum()/len(df)) * 100
print(f"Missing BMI percentage: {missing_percentage:.2f}%")

# Impute missing values with the median
df['bmi'].fillna(df['bmi'].median(), inplace=True)

# Verify that there are no more missing values
print("Remaining missing values:", df['bmi'].isnull().sum())

Missing BMI percentage: 3.93%
Remaining missing values: 0


C:\Users\shwey\AppData\Local\Temp\ipykernel_15300\949298541.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['bmi'].fillna(df['bmi'].median(), inplace=True)


# Data preprocessing

In [9]:
## Inspecting categorical variables
# Select the categorucal columns
categorical_cols = ['gender', 'ever_married', 'work_type', 'Residence_type', 'smoking_status']

# print the value counts for each categorical column
for col in categorical_cols:
    print(f"--- {col} ---")
    print(df[col].value_counts())
    print("\n")

--- gender ---
gender
Female    2994
Male      2115
Other        1
Name: count, dtype: int64


--- ever_married ---
ever_married
Yes    3353
No     1757
Name: count, dtype: int64


--- work_type ---
work_type
Private          2925
Self-employed     819
children          687
Govt_job          657
Never_worked       22
Name: count, dtype: int64


--- Residence_type ---
Residence_type
Urban    2596
Rural    2514
Name: count, dtype: int64


--- smoking_status ---
smoking_status
never smoked       1892
Unknown            1544
formerly smoked     885
smokes              789
Name: count, dtype: int64




In [11]:
## Encoding categorical variables
# Conevert categorical variables to dummy/indicator variables
df_encoded = pd.get_dummies(df, columns=['gender', 'ever_married', 'work_type', 'Residence_type', 'smoking_status'], drop_first=True, dtype=int)

# View the first rows
df_encoded.head()

,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke,gender_Male,gender_Other,ever_married_Yes,work_type_Never_worked,work_type_Private,work_type_Self-employed,work_type_children,Residence_type_Urban,smoking_status_formerly smoked,smoking_status_never smoked,smoking_status_smokes
0,67.0,0,1,228.69,36.6,1,1,0,1,0,1,0,0,1,1,0,0
1,61.0,0,0,202.21,28.1,1,0,0,1,0,0,1,0,0,0,1,0
2,80.0,0,1,105.92,32.5,1,1,0,1,0,1,0,0,0,0,1,0
3,49.0,0,0,171.23,34.4,1,0,0,1,0,1,0,0,1,0,0,1
4,79.0,1,0,174.12,24.0,1,0,0,1,0,0,1,0,0,0,1,0


In [12]:
## Separate Features(X) and Target(y)
X = df_encoded.drop(columns=['stroke'])
y = df_encoded['stroke']

# Display the shapes to confirm
print("X Shape:", X.shape)
print("y shape:", y.shape)

X Shape: (5110, 16)
y shape: (5110,)


# Model Training 

## Model 1

In [ ]:
# Train- test split

from sklearn.model_selection import train_test_split
# Split the data into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (4088, 16)
X_test shape: (1022, 16)


In [14]:
# Train the random forest model
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Initialize the model
rf_model = RandomForestClassifier(random_state=42, n_estimators=100)

# Train the model
rf_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = rf_model.predict(X_test)

In [16]:
# Evaluate the model performance
print ("Accuracy:", accuracy_score(y_test, y_pred))
print ("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.9481409001956947

Classification Report:
               precision    recall  f1-score   support

           0       0.95      1.00      0.97       972
           1       0.00      0.00      0.00        50

    accuracy                           0.95      1022
   macro avg       0.48      0.50      0.49      1022
weighted avg       0.90      0.95      0.93      1022



## Model 2

This model is not good model even though it get 94.8 % accuracy because the row 1 of the precision, recall and f1 score are 0.00. so we will make the new model.

In [19]:
## Adjust Class Weights
# Initialize the model with class_weight = 'balanced"

rf_model = RandomForestClassifier(
    random_state = 42,
    n_estimators = 100,
    class_weight = 'balanced'
)

# train the model
rf_model.fit(X_train, y_train)

# Makr predictions
y_pred = rf_model.predict(X_test)

# Evaluate
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.9500978473581213

Classification Report:
               precision    recall  f1-score   support

           0       0.95      1.00      0.97       972
           1       0.00      0.00      0.00        50

    accuracy                           0.95      1022
   macro avg       0.48      0.50      0.49      1022
weighted avg       0.90      0.95      0.93      1022



## Model 3 (With SMOTE)

The model is not still good because of the same result. so  we will use the smote .

In [23]:
# SMOTE
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score, roc_auc_score

# 1. Initialize SMOTE
smote = SMOTE(random_state=42)

# 2. Fit and apply SMOTE to the training data
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

print("Original training data shape:", X_train.shape)
print("Resampled training data shape:", X_train_resampled.shape)

# 3. Initialize and train the model
rf_model_smote = RandomForestClassifier(random_state=42, n_estimators=100)
rf_model_smote.fit(X_train_resampled, y_train_resampled)

# 4. Predict on the original test set
y_pred_smote = rf_model_smote.predict(X_test)

# 5. Evaluate
print("Accuracy:", accuracy_score(y_test, y_pred_smote))
print("\nClassification Report:\n", classification_report(y_test, y_pred_smote))

# 6. Check ROC-AUC Score
y_prob_smote = rf_model_smote.predict_proba(X_test)[:, 1]
print("ROC-AUC Score:", roc_auc_score(y_test, y_prob_smote))

Original training data shape: (4088, 16)
Resampled training data shape: (7778, 16)
Accuracy: 0.9129158512720157

Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.96      0.95       972
           1       0.09      0.08      0.08        50

    accuracy                           0.91      1022
   macro avg       0.52      0.52      0.52      1022
weighted avg       0.91      0.91      0.91      1022

ROC-AUC Score: 0.7661111111111111


## Model 4

In [24]:
## Tune the probability model

# Get probabilities for the positive class
y_prob = rf_model_smote.predict_proba(X_test)[:,1]

# Set a lower custom threshold
threshold = 0.20
y_pred_custom = (y_prob >= threshold).astype(int)

# Evaluate with the new threshold
print("Custom Threshold Accuracy:", accuracy_score(y_test, y_pred_custom))
print("\nCustom Threshold Classification Report:\n", classification_report(y_test, y_pred_custom))

Custom Threshold Accuracy: 0.8003913894324853

Custom Threshold Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.81      0.89       972
           1       0.13      0.52      0.20        50

    accuracy                           0.80      1022
   macro avg       0.55      0.67      0.54      1022
weighted avg       0.93      0.80      0.85      1022



## Model 5 (With XG Boost)

In [25]:
## Implement XGBoost
import xgboost as xgb

# 1. Calculate the scale_pos_weight ratio (non-stroke / stroke)
ratio = y_train.value_counts()[0] / y_train.value_counts()[1]

# 2. Initialize and train XG Boost
xgb_model = xgb.XGBClassifier(
    scale_pos_weight=ratio,
    random_state=42,
    n_estimators=100,
    learning_rate=0.05
)

xgb_model.fit(X_train, y_train)

# 3. Predict and evaluate
y_pred_xgb = xgb_model.predict(X_test)

print("XGBoost Classification Report:\n", classification_report(y_test, y_pred_xgb))


XGBoost Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.88      0.93       972
           1       0.20      0.58      0.30        50

    accuracy                           0.87      1022
   macro avg       0.59      0.73      0.61      1022
weighted avg       0.94      0.87      0.90      1022



## Model 6 

In [26]:
# 1. Get probabilities for the positive class
y_prob_xgb = xgb_model.predict_proba(X_test)[:, 1]

# 2. Lower the threshold
threshold = 0.18
y_pred_xgb_custom = (y_prob_xgb >= threshold).astype(int)

# 3. Evaluate the custom threshold
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
print("Custom Threshold Classification Report:\n", classification_report(y_test, y_pred_xgb_custom))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_xgb_custom))

Custom Threshold Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.67      0.80       972
           1       0.11      0.82      0.20        50

    accuracy                           0.67      1022
   macro avg       0.55      0.74      0.50      1022
weighted avg       0.94      0.67      0.77      1022


Confusion Matrix:
 [[648 324]
 [  9  41]]


# Serialize the XGBoost Model


In [28]:
import joblib
import os

# Save the trained XGBoost Model
model_path = '../model/stroke_model.pkl'
joblib.dump(xgb_model, model_path)

print(f"Model saved successfully at: {model_path}")


Model saved successfully at: ../model/stroke_model.pkl
